### PR TITLE
Proxy plugins

### PR DESCRIPTION
**Pull request by trhr**
#### Brief overview of PR changes/additions

Adds proxy plugins support (works like the Web Modules plugins, but supports serving HTTP protocols at the Django Web Root.

- settings_default.py got PROXY_PLUGIN_MODULES = ["server.conf.proxy_plugins"]
- portal.py loads above and hooks this code after web_root is created, before web_root creates Website() 
- game_template/ got a proxy_plugins.py with a passthrough function

#### Motivation for adding to Evennia

Integrations. The web plugins module doesn't support proxy service plugins. While it does take and return web_root, this is actually a different web_root than the one in portal.py; inheriting from DjangoWebsiteRoot() instead of the ReverseProxyServer in twisted.web, and while they both ultimately inherit from twisted.web.resource.Resource, the provided hooks do not allow http protocols to be served at the Django level, as the portal proxy server builds itself without any opportunity to interrupt it.

### CODE REVIEW

File: evennia/settings_default.py [hunk 1]
```diff
@@ -356,6 +356,8 @@
 MSSP_META_MODULE = "server.conf.mssp"
 # Module for web plugins.
 WEB_PLUGINS_MODULE = "server.conf.web_plugins"
+# Module for Proxy plugins
+PROXY_PLUGIN_MODULES = ["server.conf.proxy_plugins"]
```
Comment (Griatch) [inline evennia/settings_default.py]: I suggest this extra file is not needed.

File: evennia/server/portal/portal.py [hunk 2]
```diff
@@ -352,6 +357,9 @@ def shutdown(self, _reactor_stopping=False, _stop_server=False):
                 web_root.putChild(b"webclientdata", ajax_webclient)
                 webclientstr = "webclient (ajax only)"
 
+                for plugin_module in PROXY_PLUGIN_MODULES:
```
[the remark this answers is not in the document]
Comment (trhr) [reply evennia/server/portal/portal.py]: You're right, it should be outdented, but I think before 390 is better than after 391. Users should not be able to block the website from creating an externally-accessible portal (web_root.is_portal = True), and that portal should be a Website().

File: evennia/game_template/server/conf/proxy_plugins.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+"""
+Proxy plugin hooks.
```
[the remark this answers is not in the document]
Comment (trhr) [reply evennia/game_template/server/conf/proxy_plugins.py]: Agree on both this and the settings.py change.

File: evennia/server/portal/portal.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Griatch) [reply evennia/server/portal/portal.py]: Maybe, but if they do, it'd be a very deliberate choice I'd think ... who knows what shenanigans people want to use the plugin for. ;)

[the remark this answers is not in the document]
Comment (trhr) [reply evennia/server/portal/portal.py]: As long as putChild inherits the amp protocol details (this was the main sticking point with working with new protocols in web_plugins.py), it seems ok. That inheritance would happen in the upcoming internet.TCPServer call anyway. So, why not!

**Comment by trhr** [comment]
Whoops, merged the Django3 pull req when I tried to test in my dev environment. I think I reverted all those.

**Comment by trhr** [comment]
This is too gitvanced for me, I'm not sure how the two simultaneous pull requests work or why this is conflicted; there should be no changes in urls.py on this PR...

**Comment by Griatch** [comment]
@trhr Not sure what you did here, but your PR reverts a lot of changes in master. You need to clean this PR up to only include the changes you made so it can be applied cleanly. I'd suggest regularly sync your work-branch with `master` and resolve any conflicts so as to keep the merge clean.

**Pull request closed without merging**